021-server-side-rendering/021-server-side-rendering.md

```diff
@@ -0,0 +1,304 @@
+
+A popular hacky workaround for this involves making a manual fetch request, bypassing the SDK and parsing the cookie from the response headers.
+
+```js
+import { parse } from "set-cookie-parser";
+
+const response = await fetch(
+  `https://cloud.appwrite.io/v1/account/sessions/anonymous`,
+  {
+    method: "POST",
+    headers: {
+      "x-appwrite-project": "PROJECT_ID",
+    },
+  }
+);
+
+const cookies = response.headers.get("set-cookie") || "";
+const sessionCookie = parse(cookies).find((cookie) =>
+  cookie.name.startsWith("a_session")
+);
+const sessionToken = sessionCookie?.value;
+```
+
+The SDK for flutter has a similiar problem, but bypasses it by adding and interceptor to the Client requests. The interceptor sets cookies (in the apps 'cookie jar') when receiving set-cookie responses, and retrieves the cookie when making requests. For server side SDKs, this is not possible, as there isn't an equivalent method persistent storage.
+
+**Problem #2: Accessing OAuth2 sessions**
+
+The current oauth2 flow looks something like this:
+
+![CSR OAuth2 Flow Sequence Diagram](csr-oauth-flow.png)
+
+1. Browser makes a GET request to the Appwrite API.
+2. Appwrite API redirects the browser to the OAuth2 provider.
+3. User authenticates with the OAuth2 provider.
+4. OAuth2 provider redirects the browser back to the Appwrite API.
+5. Appwrite API sets the session cookie on the Appwrite domain.
+6. Appwrite API redirects the browser back to the client application.
+
+This is incompatible with SSR applications, because the session cookie is set on the Appwrite domain, and not the SSR domain.
+
+Interestingly, there is an undocumented workaround for this. If the developer sets the success parameter to `{SSR_DOMAIN}/auth/oauth2/success` the Appwrite API will include the session cookie as a query parameter in the redirect URL. [Source here](https://github.com/appwrite/appwrite/blob/3f3d518f3664bcab281ee00b45dd2f2d387ffc72/app/controllers/api/account.php#L870). This method is vunerable to session spying attacks, as the session token is exposed in the URL, and could be used to hijack the session.
+
+**Problem #3: Using session tokens with the SDK**
+
+Now, you've got a session token, you want to make authenticated requests. For example, getting the user's account details. To set a session token on the server-side, we have to use the `X-Fallback-Cookies` header
+
+```js
+import { Client, Account } from "appwrite";
+
+const client = new Client();
+client.setEndpoint("https://cloud.appwrite.io/v1");
+client.setProject("PROJECT_ID");
+
+const serialisedCookies = JSON.stringify({
+  a_session_PROJECT_ID: sessionToken,
+});
+
+client.headers["X-Fallback-Cookies"] = serialisedCookies;
+
+const account = new Account(client);
+const currentUser = account.get();
+```
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+**Problem #1: Solution A - Return session token in response body**
```

Comment:
I like solution A, can't see why not.
Such token should be hidden by default, and only exposed when creating session.

Reply:
What do you mean by `such token should be hidden by default`?

Reply:
You had 2 solutions. One to have it in response body, one to set it as response header.

I like setting it in response body. But if in future you do `getSession`, you should no longer get secret. You should only get it once when you first create it.